UrlConnectionHttpClient doesn't honor TRUST_ALL_CERTIFICATES

<!--- Provide a general summary of the issue in the Title above -->

## Expected Behavior
When setting `TRUST_ALL_CERTIFICATES` to true, using `UrlConnectionHttpClient` should work for local https server

## Current Behavior
Exception thrown
```
Caused by: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
```

## Steps to Reproduce (for bugs)
Using UrlConnectionHttpClient to send request to local https server

## Context
<!--- How has this issue affected you? What are you trying to accomplish? -->
<!--- Providing context helps us come up with a solution that is most useful in the real world -->

## Your Environment
<!--- Include as many relevant details about the environment where the bug was discovered -->
* AWS Java SDK version used:  N/A
* JDK version used: N/A
* Operating System and version: N/A

--- a/.changes/next-release/bugfix-URLConnectionHttpClient-8ae3a85.json
+++ b/.changes/next-release/bugfix-URLConnectionHttpClient-8ae3a85.json
@@ -1,0 +1,5 @@
+{
+    "category": "URL Connection Http Client", 
+    "type": "bugfix", 
+    "description": "Bypass ssl validations when `TRUST_ALL_CERTIFICATES` is set to true."
+}

--- a/http-clients/url-connection-client/src/main/java/software/amazon/awssdk/http/urlconnection/UrlConnectionHttpClient.java
+++ b/http-clients/url-connection-client/src/main/java/software/amazon/awssdk/http/urlconnection/UrlConnectionHttpClient.java
@@ -26,10 +26,20 @@ import java.io.IOException;
 import java.io.InputStream;
 import java.net.HttpURLConnection;
 import java.net.URI;
+import java.security.KeyManagementException;
+import java.security.NoSuchAlgorithmException;
+import java.security.cert.CertificateException;
+import java.security.cert.X509Certificate;
 import java.time.Duration;
 import java.util.List;
 import java.util.Map;
 import java.util.stream.Collectors;
+import javax.net.ssl.HostnameVerifier;
+import javax.net.ssl.HttpsURLConnection;
+import javax.net.ssl.SSLContext;
+import javax.net.ssl.SSLSession;
+import javax.net.ssl.TrustManager;
+import javax.net.ssl.X509TrustManager;
 import software.amazon.awssdk.annotations.SdkPublicApi;
 import software.amazon.awssdk.http.AbortableInputStream;
 import software.amazon.awssdk.http.ExecutableHttpRequest;
@@ -120,10 +130,34 @@ public final class UrlConnectionHttpClient implements SdkHttpClient {
     }
 
     private HttpURLConnection createDefaultConnection(URI uri) {
+
+        if (options.get(SdkHttpConfigurationOption.TRUST_ALL_CERTIFICATES)) {
+            trustAllCertificates();
+        }
+
         HttpURLConnection connection = invokeSafely(() -> (HttpURLConnection) uri.toURL().openConnection());
         connection.setConnectTimeout(saturatedCast(options.get(CONNECTION_TIMEOUT).toMillis()));
         connection.setReadTimeout(saturatedCast(options.get(READ_TIMEOUT).toMillis()));
         return connection;
+    }
+
+    /**
+     * Should only be used in testing
+     */
+    private static void trustAllCertificates() {
+        HttpsURLConnection.setDefaultHostnameVerifier(NoOpHostNameVerifier.INSTANCE);
+
+        TrustManager[] trustManagers = new TrustManager[]{TrustAllManager.INSTANCE};
+        SSLContext context;
+
+        try {
+            context = SSLContext.getInstance("TLS");
+            context.init(null, trustManagers, null);
+        } catch (NoSuchAlgorithmException | KeyManagementException ex) {
+            throw new RuntimeException(ex.getMessage(), ex);
+        }
+
+        HttpsURLConnection.setDefaultSSLSocketFactory(context.getSocketFactory());
     }
 
     private static class RequestCallable implements ExecutableHttpRequest {
@@ -249,6 +283,39 @@ public final class UrlConnectionHttpClient implements SdkHttpClient {
                                                               .merge(serviceDefaults)
                                                               .merge(SdkHttpConfigurationOption.GLOBAL_HTTP_DEFAULTS),
                                                null);
+        }
+    }
+
+    private static class NoOpHostNameVerifier implements HostnameVerifier {
+
+        static final NoOpHostNameVerifier INSTANCE = new NoOpHostNameVerifier();
+
+        @Override
+        public boolean verify(String s, SSLSession sslSession) {
+            return true;
+        }
+    }
+
+    /**
+     * Insecure trust manager to trust all certs. Should only be used for testing.
+     */
+    private static class TrustAllManager implements X509TrustManager {
+
+        private static final TrustAllManager INSTANCE = new TrustAllManager();
+
+        @Override
+        public void checkClientTrusted(X509Certificate[] x509Certificates, String s) throws CertificateException {
+            // no op
+        }
+
+        @Override
+        public void checkServerTrusted(X509Certificate[] x509Certificates, String s) throws CertificateException {
+            // no op
+        }
+
+        @Override
+        public X509Certificate[] getAcceptedIssuers() {
+            return new X509Certificate[0];
         }
     }
 }

--- a/http-clients/url-connection-client/src/test/java/software/amazon/awssdk/http/urlconnection/UrlConnectionHttpClientWireMockTest.java
+++ b/http-clients/url-connection-client/src/test/java/software/amazon/awssdk/http/urlconnection/UrlConnectionHttpClientWireMockTest.java
@@ -14,12 +14,37 @@
  */
 package software.amazon.awssdk.http.urlconnection;
 
+import static software.amazon.awssdk.http.SdkHttpConfigurationOption.TRUST_ALL_CERTIFICATES;
+
+import java.net.HttpURLConnection;
+import javax.net.ssl.HttpsURLConnection;
+import javax.net.ssl.SSLSocketFactory;
+import org.junit.After;
+import org.junit.Test;
 import software.amazon.awssdk.http.SdkHttpClient;
 import software.amazon.awssdk.http.SdkHttpClientTestSuite;
+import software.amazon.awssdk.utils.AttributeMap;
 
 public final class UrlConnectionHttpClientWireMockTest extends SdkHttpClientTestSuite {
+
     @Override
     protected SdkHttpClient createSdkHttpClient(SdkHttpClientOptions options) {
         return UrlConnectionHttpClient.create();
+    }
+
+    @After
+    public void reset() {
+        HttpsURLConnection.setDefaultSSLSocketFactory((SSLSocketFactory) SSLSocketFactory.getDefault());
+    }
+
+    @Test
+    public void trustAllCertificates_shouldWork() throws Exception {
+        try (SdkHttpClient client = UrlConnectionHttpClient.builder()
+                                                           .buildWithDefaults(AttributeMap.builder()
+                                                                                          .put(TRUST_ALL_CERTIFICATES,
+                                                                                               Boolean.TRUE)
+                                                                                          .build())) {
+            testForResponseCodeUsingHttps(client, HttpURLConnection.HTTP_OK);
+        }
     }
 }

